### ISSUE TITLE
Passing locals as arguments directly

### ISSUE THREAD

**Issue by katyo** [MaulingMonkey/jni-bindgen#62]
Currently the arguments defined using types directly.

As example, the `Throwable` constructor, which gets string:
```rust
pub fn new_String<'env>(
    __jni_env: &'env Env,
    arg0: impl Into<Option<&'env String>>
) -> Result<Local<'env, Throwable>, Local<'env, Throwable>>
```

Possibly more convenient way is something like:
```rust
pub fn new_String<'env, Arg0: AsRef<String>>(
    __jni_env: &'env Env,
    arg0: impl Into<Option<Arg0>>
) -> Result<Local<'env, Throwable>, Local<'env, Throwable>>
```

But because local hasn't AsRef impl, so maybe like follow:
```rust
pub fn new_String<'env, Arg0: Deref<Target = &String>>(
    __jni_env: &'env Env,
    arg0: impl Into<Option<Arg0>>
) -> Result<Local<'env, Throwable>, Local<'env, Throwable>>
```

**Comment by katyo** [comment]
I think this is harmless to add AsRef impl to `Local`.

**Comment by MaulingMonkey** [comment]
IIRC the reason I haven't done this yet is because of ambiguity errors that crop up when trying to be generic over two levels of indirection (e.g. passing `None` fails IIRC?).  It'd be nice to improve this to take something generic like AsRef, but I want to start fleshing out some good unit test / example coverage before trying to tackle this.

**Comment by katyo** [comment]
Oh this is epic problem in rust, so I take pain avoid more then one level of indirection.
Maybe something as:
```rust
pub fn new_String<'env, Arg0: AsRef<String>>(
    __jni_env: &'env Env,
    arg0: Option<Arg0>,
) -> Result<Local<'env, Throwable>, Local<'env, Throwable>>
```
Despite the fact that it requires `Some` for non null values, but it is explicit and not so big inconvenient.

**Comment by MaulingMonkey** [comment]
Another option might be to generate a bunch of `impl Into<Option<&'env java::lang::Object>> for ...`s, where `...` might include `Local<&String>`, `Option<Local<&String>>`, etc.

However, I'm already hitting docs.rs limits on crate size, having horrible build times, etc. so I want to see what I can do to fix that some first.  I've been doing various experiments with proc macros, and I'm considering getting rid of the `jni-android-sys` crate and its thousands of autogenerated features, in favor of generating just a subset of the bindings needed for just your local crate.

**Comment by katyo** [comment]
Yes, compile time code generation is a right thing especially in case of bindings.
